### ISSUE TITLE
Document and improve hooks

### ISSUE THREAD

**Issue by RDIL** [thepeacockproject/Peacock#21]
- [ ] Document the hooks
- [ ] Maybe improve the API?
- [ ] Add new hooks where plugin makers need them

**Comment by AnthonyFuller** [comment]
Areas for hook improvements:
- Challenges
- CPD (Maybe allow modders to do some cool stuff with this? May not need hooks.)
- Mastery
- Scoring(?)
- Multiplayer(?)

**Comment by RDIL** [comment]
Additionally, add ability to untap, add priorities, etc

**Comment by AnthonyFuller** [comment]
Can we not untap? Thought we could.

**Comment by RDIL** [comment]
I don't think we can officially right now, you can always hack around it e.g.

```ts
let a = true

hooks.example.tap("MyPlugin", () => {
    if (!a) return
    console.log("still tapped")
    a = false
    console.log("untapped for future use")
})
```

**Comment by AnthonyFuller** [comment]
Ah okay, that's fair. Should we still use tapable or move away from that? I vaguely recall you complaining about it in the past.

**Comment by RDIL** [comment]
We already migrated off tapable because it's codegen at runtime approach could be buggy, [this](https://github.com/thepeacockproject/Peacock/blob/v7/components/hooksImpl.ts) is what we're using rn

**Comment by AnthonyFuller** [comment]
Ah right I forgot about that change, mostly because we kept the API design the same.
